File: src/stores/writeGate.ts
import { defineStore } from "pinia";
import { ref, computed } from "vue";
import { invoke } from "@tauri-apps/api/core";

export interface PendingWriteRecord {
  id: string;
  subsystem: string; // "memory" | "skills"
  action: string;
  summary: string;
  origin: string;
  created_at: number | null;
  payload: Record<string, any>;
}

export interface PendingWriteReviewNote {
  type: string;
  target_label?: string;
  skill_name?: string;
}

export interface PendingWriteReview {
  subsystem: string;
  target_label: string;
  language: string;
  current: string;
  proposed: string;
  diff: string;
  requested_old_string?: string;
  payload_text?: string;
  notes: PendingWriteReviewNote[];
}

export interface PendingWritesResponse {
  records: PendingWriteRecord[];
  counts: Record<string, number>;
  supported?: boolean;
}

export const useWriteGateStore = defineStore("writeGate", () => {
  const records = ref<PendingWriteRecord[]>([]);
  const counts = ref<Record<string, number>>({});
  const supported = ref(true);
  const loading = ref(false);
  const activeAction = ref(""); // "id:diff" | "id:approve" | "id:reject"
  const expandedReviews = ref<Record<string, PendingWriteReview>>({});

  const pendingCount = computed(() => records.value.length);

  function recordKey(r: PendingWriteRecord): string {
    return `${r.subsystem}:${r.id}`;
  }

  async function fetchPending() {
    loading.value = true;
    try {
      const data: PendingWritesResponse = await invoke("write_gate_list");
      supported.value = data.supported !== false;
      records.value = supported.value ? data.records || [] : [];
      counts.value = data.counts || {};
    } catch (e: any) {
      console.error("Failed to fetch pending writes:", e);
      supported.value = false;
    } finally {
      loading.value = false;
    }
  }

  async function fetchReview(record: PendingWriteRecord): Promise<PendingWriteReview> {
    const key = recordKey(record);
    activeAction.value = `${key}:diff`;
    try {
      const review: PendingWriteReview = await invoke("write_gate_diff", {
        subsystem: record.subsystem,
        id: record.id,
      });
      expandedReviews.value = { ...expandedReviews.value, [key]: review };
      return review;
    } finally {
      activeAction.value = "";
    }
  }

  function toggleReview(record: PendingWriteRecord) {
    const key = recordKey(record);
    if (expandedReviews.value[key]) {
      const next = { ...expandedReviews.value };
      delete next[key];
      expandedReviews.value = next;
    } else {
      fetchReview(record);
    }
  }

  async function approve(record: PendingWriteRecord) {
    const key = recordKey(record);
    activeAction.value = `${key}:approve`;
    try {
      await invoke("write_gate_approve", {
        subsystem: record.subsystem,
        id: record.id,
      });
      // Remove from expanded
      const next = { ...expandedReviews.value };
      delete next[key];
      expandedReviews.value = next;
      await fetchPending();
    } finally {
      activeAction.value = "";
    }
  }

  async function reject(record: PendingWriteRecord) {
    const key = recordKey(record);
    activeAction.value = `${key}:reject`;
    try {
      await invoke("write_gate_reject", {
        subsystem: record.subsystem,
        id: record.id,
      });
      const next = { ...expandedReviews.value };
      delete next[key];
      expandedReviews.value = next;
      await fetchPending();
    } finally {
      activeAction.value = "";
    }
  }

  function formatTime(value: number | null): string {
    if (!value) return "";
    return new Date(value * 1000).toLocaleString([], {
      month: "short",
      day: "numeric",
      hour: "2-digit",
      minute: "2-digit",
    });
  }

  return {
    records,
    counts,
    supported,
    loading,
    activeAction,
    expandedReviews,
    pendingCount,
    recordKey,
    fetchPending,
    fetchReview,
    toggleReview,
    approve,
    reject,
    formatTime,
  };
});

File: src-tauri/src/write_gate.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ═══════════════════════════════════════════
// Types — mirrors Hermes Studio write-gate API
// ═══════════════════════════════════════════

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PendingWriteRecord {
    pub id: String,
    pub subsystem: String, // "memory" | "skills"
    pub action: String,
    pub summary: String,
    pub origin: String,
    pub created_at: Option<f64>,
    #[serde(default)]
    pub payload: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PendingWriteReviewNote {
    #[serde(rename = "type")]
    pub note_type: String,
    #[serde(default)]
    pub target_label: Option<String>,
    #[serde(default)]
    pub skill_name: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PendingWriteReview {
    pub subsystem: String,
    pub target_label: String,
    pub language: String,
    pub current: String,
    pub proposed: String,
    pub diff: String,
    #[serde(default)]
    pub requested_old_string: Option<String>,
    #[serde(default)]
    pub payload_text: Option<String>,
    #[serde(default)]
    pub notes: Vec<PendingWriteReviewNote>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PendingWritesResponse {
    pub records: Vec<PendingWriteRecord>,
    pub counts: HashMap<String, u32>,
    #[serde(default)]
    pub supported: Option<bool>,
}

// ═══════════════════════════════════════════
// Hermes Studio API helpers
// ═══════════════════════════════════════════

fn hermes_base_url() -> String {
    std::env::var("HERMES_STUDIO_URL")
        .unwrap_or_else(|_| "http://localhost:8648".to_string())
}

fn hermes_client() -> reqwest::blocking::Client {
    reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(15))
        .build()
        .unwrap_or_default()
}

// ═══════════════════════════════════════════
// Tauri Commands
// ═══════════════════════════════════════════

#[tauri::command]
pub async fn write_gate_list() -> Result<PendingWritesResponse, String> {
    let url = format!("{}/api/hermes/write-gate/pending", hermes_base_url());
    let client = hermes_client();
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    let data: PendingWritesResponse = resp.json().map_err(|e| e.to_string())?;
    Ok(data)
}

#[tauri::command]
pub async fn write_gate_diff(subsystem: String, id: String) -> Result<PendingWriteReview, String> {
    let url = format!(
        "{}/api/hermes/write-gate/pending/{}/diff",
        hermes_base_url(), id
    );
    let client = hermes_client();
    let resp = client
        .get(&url)
        .query(&[("subsystem", &subsystem)])
        .send()
        .map_err(|e| e.to_string())?;

    // The API returns { diff, review? }
    let raw: serde_json::Value = resp.json().map_err(|e| e.to_string())?;

    if let Some(review) = raw.get("review") {
        let r: PendingWriteReview =
            serde_json::from_value(review.clone()).map_err(|e| e.to_string())?;
        return Ok(r);
    }

    // Fallback: construct review from diff
    let diff_text = raw
        .get("diff")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    Ok(PendingWriteReview {
        subsystem: subsystem.clone(),
        target_label: subsystem,
        language: String::new(),
        current: String::new(),
        proposed: diff_text.clone(),
        diff: diff_text,
        requested_old_string: None,
        payload_text: None,
        notes: vec![],
    })
}

#[tauri::command]
pub async fn write_gate_approve(subsystem: String, id: String) -> Result<String, String> {
    let url = format!(
        "{}/api/hermes/write-gate/pending/{}/approve",
        hermes_base_url(), id
    );
    let client = hermes_client();
    let resp = client
        .post(&url)
        .json(&serde_json::json!({ "subsystem": subsystem }))
        .send()
        .map_err(|e| e.to_string())?;
    let body: serde_json::Value = resp.json().map_err(|e| e.to_string())?;
    Ok(body
        .get("output")
        .and_then(|v| v.as_str())
        .unwrap_or("Approved")
        .to_string())
}

#[tauri::command]
pub async fn write_gate_reject(subsystem: String, id: String) -> Result<String, String> {
    let url = format!(
        "{}/api/hermes/write-gate/pending/{}/reject",
        hermes_base_url(), id
    );
    let client = hermes_client();
    let resp = client
        .post(&url)
        .json(&serde_json::json!({ "subsystem": subsystem }))
        .send()
        .map_err(|e| e.to_string())?;
    let body: serde_json::Value = resp.json().map_err(|e| e.to_string())?;
    Ok(body
        .get("output")
        .and_then(|v| v.as_str())
        .unwrap_or("Rejected")
        .to_string())
}
